tuna/machine.py

```diff
@@ -283,31 +306,55 @@ def get_properties(self):
 
     return self.cpus, self.gpus
 
-  def write_file(self, contents, filename=None, is_temp=False):
+  def write_file(self, contents: bytes, filename: Union[str, bytes, 'os.PathLike[str]',\
+  None] = None, is_temp: bool = False) -> Union[str, bytes, 'os.PathLike[str]', None]:
     """
     Write a file to this machine containing contents
     """
+    cnx: Connection
+    ftp: Optional[paramiko.sftp_client.SFTPClient] = None
+    fout: SFTPFile
+    fout_2: IO[Any]
+    t_file: bool = False
+
+    t_filename: Union[str, 'os.PathLike[str]']
     if is_temp:
       assert filename is None
-      _, filename = tempfile.mkstemp()
+      _, t_filename = tempfile.mkstemp()
+      t_file = True
     else:
       assert filename is not None
-    if self.local_machine:  # pylint: disable=no-member ; false alarm
-      with open(filename, 'wb') as fout:
-        fout.write(contents)
-        fout.flush()
+
+    if self.local_machine and filename is not None:  # pylint: disable=no-member ; false alarm
+      with open(filename, 'wb') as fout_2:
+        fout_2.write(contents)
+        fout_2.flush()
     else:
       cnx = self.connect()
-      ftp = cnx.ssh.open_sftp()
-      with ftp.open(filename, 'wb') as fout:
-        fout.write(contents)
-        fout.flush()
+      if cnx is not None and cnx.ssh is not None:
+        ftp = cnx.ssh.open_sftp()
+      if ftp is not None:
+        with ftp.open(filename, 'wb') as fout:  #type: ignore
+          fout.write(contents)
+          fout.flush()
```

Comment:
this is why you get the error, this code is not reached, you have changed the way the code works here, please undo